EIPS/eip-2565.md

```diff
@@ -4,34 +4,48 @@ title: Repricing of the EIP-198 ModExp precompile
-review-period-end: 2020-07-31
+review-period-end: 2020-10-09
 type: Standards Track
 category: Core
 created: 2020-03-20
 requires: 198
 ---
 
 ## Simple Summary
-The [EIP-198](https://eips.ethereum.org/EIPS/eip-198) ‘big integer modular exponentiation’, or `ModExp`, precompile is currently overpriced. Re-pricing this precompile will enable more cost efficient verification of RSA signatures, verifiable delay functions (VDFs), primality checks, and more.
+This EIP reprices the [EIP-198](./eip-198.md) ‘big integer modular exponentiation’, or `ModExp`, precompile.
 
 ## Abstract
-After benchmarking the ModExp precompile, we discovered that it is ‘overpriced’ relative to other precompiles. We also discovered that the current gas pricing formula could be improved to better estimate the computational complexity of various ModExp input variables. To improve the gas cost pricing for this precompile the following options are available:
-
-1. Modifying the gas pricing formula to better reflect the computational complexity of ModExp operations
-2. Changing the value of the `GQUADDIVISOR` parameter in the ModExp pricing formula to bring its costs more in-line with other precompiles
-3. Improving the underlying libraries beneath the ModExp Precompile
-4. Any combination of (1), (2), and (3)
-
-We recommend **Option (1) and (2)** which provides a large practical improvement to gas estimation while keeping implementation complexity low. Option (3) can be implemented at a later point to improve the gas pricing even further.
+To accurately reflect the real world operational cost of the `ModExp` precompile, this EIP specifies an algorithm for calculating the gas cost. This algorithm approximates the multiplication complexity cost and multiplies that by an approximation of the iterations required to execute the exponentiation.
 
 ## Motivation
 Modular exponentiation is a foundational arithmetic operation for many cryptographic functions including signatures, VDFs, SNARKs, accumulators, and more. Unfortunately, the ModExp precompile is currently over-priced, making these operations inefficient and expensive. By reducing the cost of this precompile, these cryptographic functions become more practical, enabling improved security, stronger randomness (VDFs), and more.
 
 ## Specification
-The current gas pricing formula is defined in EIP-198. This formula divides a ‘computational complexity’ function by a ‘gas conversion’ parameter called `GQUADDIVISOR` to arrive at a gas cost: `floor(mult_complexity(x)/GQUADDIVISOR)`
+As of `FORK_BLOCK_NUMBER`, the gas cost of calling the precompile at address `0x0000000000000000000000000000000000000005` will be calculated as follows:
+```
+def calculate_multiplication_complexity(base_length, modulus_length):
+    max_length = max(base_length, modulus_length)
+    words = math.ceil(max_length / 8)
+    return words**2
+
+def calculate_iteration_count(exponent_length, exponent): # same as ADUSTED_EXPONENT_LENGTH in EIP-198
```

Comment:
```suggestion
def calculate_iteration_count(exponent_length, exponent):
```
Since this EIP stands on its own, no need to reference 198 here as it could lead to confusion (it confused me at least).

Reply:
@MicahZoltu I added this to specifically to help implementers who are migrating from EIP-198 to EIP-2565, which would be the case at least for Ethereum mainnet. I'm happy to delete if it is confusing. @holiman previously I had structured this EIP as changes that should be made on top of EIP-198. We've since changed it to be more of a 'stand alone' EIP that could be adopted in lieu of EIP-198's gas pricing. It would be great to hear your perspective on if these sort of notes are useful for implementers or only add confusion.